bigquery/google/cloud/bigquery/_helpers.py

```diff
@@ -328,10 +328,15 @@ def __init__(self, udf_type, value):
         self.value = value
 
     def __eq__(self, other):
+        if not isinstance(other, UDFResource):
+            return NotImplemented
         return(
             self.udf_type == other.udf_type and
             self.value == other.value)
 
+    def __ne__(self, other):
+        return not self == other
```

Comment:
The reason why: The entire "right thing" happens with the `==` operator. So, `self.__eq__(other)` gets called, _but_ unlike if we returned `not self.__eq__(other)` directly, if it returns `NotImplemented`, then the other object is asked and it is completely resolved to the boolean, which is then inverted.

Therefore, while `__ne__` never returns `NotImplemented`, it does correctly resolve the one that `__eq__` returns.